File: internal/api/idempotency.go
package api

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"io"
	"net/http"
	"strings"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/school-erp/erp/internal/httpx"
)

/* SENDING THE SAME WRITE TWICE MUST NOT DO IT TWICE.

   This is the half of offline support that lives on the server. The client
   holds writes it could not send and sends them when the signal returns, and
   that is only safe because of what happens here.

   The problem it solves is not "the request failed". It is that a client
   cannot tell a request that never arrived from one that arrived, was
   executed, and whose answer was lost on the way back. Those are the same
   event from where the client is standing: no response. Retrying is wrong in
   the second case and required in the first, and the client has no way to
   know which it is in.

   So the client names the attempt instead. `Idempotency-Key` is minted once,
   when somebody presses the button, and every retry of that press carries the
   same value. The first request to arrive under a key does the work and its
   answer is stored; every later arrival is answered from the store and the
   handler never runs. A dropped response now costs a duplicate REQUEST, which
   is free, instead of a duplicate WRITE, which is a family charged twice.

   Deliberately opt-in per request rather than applied to every mutation. A key
   the client did not mint is a key the client cannot resend, and inventing one
   here from the method and path would make two genuinely separate presses of
   "add payment" -- same route, same body, two real payments -- collapse into
   one. Only the caller knows whether two identical requests are one intent or
   two, so only the caller may say so. */

// Requests that carry no key, and reads, pass through untouched.
func (s *Server) Idempotent(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		key := strings.TrimSpace(r.Header.Get("Idempotency-Key"))
		switch {
		case key == "",
			len(key) > 200,
			r.Method == http.MethodGet, r.Method == http.MethodHead, r.Method == http.MethodOptions:
			next.ServeHTTP(w, r)
			return
		}

		/* JSON only, and that is a safety rule rather than a convenience.
		   The body has to be read to hash it, and reading a multipart upload
		   into memory to hash it would cap every file the product accepts at
		   whatever limit this middleware picked -- a bulk-import sheet or a
		   student photograph would start failing for a reason nobody could
		   see from the screen. File uploads are also not the case this
		   exists for: nobody re-sends a photograph from a queue on a bus. */
		if ct := r.Header.Get("Content-Type"); ct != "" &&
			!strings.HasPrefix(strings.ToLower(ct), "application/json") {
			next.ServeHTTP(w, r)
			return
		}

		id := httpx.IdentityFrom(r.Context())
		/* A platform operator not working inside a school has no tenant to
		   scope the receipt to, and the table is keyed by one. They are also
		   not the caller this exists for -- nobody administers the platform
		   from a bus -- so the honest thing is to let the request through
		   unprotected rather than store a row under a zero institution that
		   every other operator would then collide with. */
		if id == nil || id.InstitutionID == uuid.Nil {
			next.ServeHTTP(w, r)
			return
		}

		/* The body has to be read to hash it, and reading it consumes it.
		   Put it back before anything downstream looks. */
		body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, 8<<20))
		if err != nil {
			httpx.BadRequest(w, r, "could not read the request body")
			return
		}
		r.Body = io.NopCloser(bytes.NewReader(body))
		sum := sha256.Sum256(body)
		hash := hex.EncodeToString(sum[:])

		scope := tenantScope(id)
		var (
			claimed bool
			prevSt  *int
			prevBod []byte
			prevM   string
			prevP   string
			prevH   string
		)
		err = s.DB.InTenant(r.Context(), scope, func(tx pgx.Tx) error {
			/* Claim and read in one statement. ON CONFLICT DO NOTHING returns
			   no row when the key is already there, which is exactly the
			   signal needed -- and it is atomic, so two requests racing under
			   one key cannot both believe they are the first. */
			err := tx.QueryRow(r.Context(), `
                INSERT INTO idempotency_keys
                       (institution_id, key, user_id, method, path, request_hash)
                VALUES ($1, $2, $3, $4, $5, $6)
                ON CONFLICT (institution_id, key) DO NOTHING
                RETURNING true`,
				id.InstitutionID, key, id.UserID, r.Method, r.URL.Path, hash,
			).Scan(&claimed)
			if err == nil {
				return nil
			}
			if !errors.Is(err, pgx.ErrNoRows) {
				return err
			}
			return tx.QueryRow(r.Context(), `
                SELECT status_code, response_body, method, path, request_hash
                  FROM idempotency_keys
                 WHERE institution_id = $1 AND key = $2 AND user_id = $3`,
				id.InstitutionID, key, id.UserID,
			).Scan(&prevSt, &prevBod, &prevM, &prevP, &prevH)
		})
		switch {
		case err == nil:
		case errors.Is(err, pgx.ErrNoRows):
			/* The key exists for somebody else. Two people share the
			   staffroom laptop; answering this one from the other one's
			   receipt would hand over their data. */
			httpx.Error(w, r, http.StatusConflict, "idempotency_key_reused",
				"That request reference belongs to a different sign-in. Try the action again.")
			return
		default:
			httpx.Error(w, r, http.StatusInternalServerError, "idempotency_failed",
				"Could not check whether this request had already been sent.")
			return
		}

		if !claimed {
			// A key reused for a different request is a client bug, and
			// answering it from the wrong receipt would hide it.
			if prevM != r.Method || prevP != r.URL.Path || prevH != hash {
				httpx.Error(w, r, http.StatusConflict, "idempotency_key_conflict",
					"That request reference was already used for a different request.")
				return
			}
			if prevSt == nil {
				/* Claimed, not finished. The first copy is still running;
				   starting the work again beside it is the duplicate this
				   whole file exists to prevent. 409 rather than 425 because
				   the client outbox already treats 4xx as "stop and show me",
				   and this is genuinely a conflict with a request in flight. */
				httpx.Error(w, r, http.StatusConflict, "idempotency_in_flight",
					"That request is still being processed. It will not be sent twice.")
				return
			}
			w.Header().Set("Content-Type", "application/json")
			w.Header().Set("Idempotent-Replay", "true")
			w.WriteHeader(*prevSt)
			_, _ = w.Write(prevBod)
			return
		}

		rec := &recorder{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(rec, r)

		/* Only a settled answer is worth keeping. A 5xx is the server saying
		   it does not know what happened, and storing that would turn one bad
		   minute into a permanent answer: every retry of a write that might
		   yet have succeeded would be told "500" forever, and the row could
		   never be re-attempted. Releasing the claim lets the next retry try
		   for real. 4xx IS settled -- the request was wrong and will be wrong
		   again -- so it is stored like any other outcome. */
		if rec.status >= 500 {
			_ = s.DB.InTenant(r.Context(), scope, func(tx pgx.Tx) error {
				_, err := tx.Exec(r.Context(),
					`DELETE FROM idempotency_keys WHERE institution_id = $1 AND key = $2`,
					id.InstitutionID, key)
				return err
			})
			return
		}
		_ = s.DB.InTenant(r.Context(), scope, func(tx pgx.Tx) error {
			_, err := tx.Exec(r.Context(), `
                UPDATE idempotency_keys
                   SET status_code = $3, response_body = $4, completed_at = now()
                 WHERE institution_id = $1 AND key = $2`,
				id.InstitutionID, key, rec.status, rec.buf.Bytes())
			return err
		})
	})
}

/*
Keeps a copy of what the handler said.

	Writes straight through as well as buffering, so a slow handler still
	streams and nothing about the live response changes shape. The buffer is
	only read afterwards, to be stored.
*/
type recorder struct {
	http.ResponseWriter
	status int
	buf    bytes.Buffer
	wrote  bool
}

func (rec *recorder) WriteHeader(code int) {
	if rec.wrote {
		return
	}
	rec.wrote = true
	rec.status = code
	rec.ResponseWriter.WriteHeader(code)
}

func (rec *recorder) Write(p []byte) (int, error) {
	if !rec.wrote {
		rec.WriteHeader(http.StatusOK)
	}
	rec.buf.Write(p)
	return rec.ResponseWriter.Write(p)
}

File: worker/src/idempotency.ts
import type { Identity } from './identity'
import { HttpError, errorResponse, now } from './http'

/* Port of internal/api/idempotency.go (Server.Idempotent). A write carrying an
   Idempotency-Key runs once per (school, key); a repeat is answered from the
   stored receipt with `Idempotent-Replay: true`. Keyless requests, reads,
   non-JSON bodies and callers with no school pass through untouched.

   D1 has no interactive transactions or row locks: the claim is a single
   INSERT ... ON CONFLICT DO NOTHING, whose `changes` says whether this request
   is the first. The receipt is settled (or released on a 5xx) afterwards. */
export async function idempotent(
  req: Request, id: Identity, db: () => D1Database, next: (req: Request) => Promise<Response>,
): Promise<Response> {
  const key = (req.headers.get('Idempotency-Key') ?? '').trim()
  const m = req.method
  if (key === '' || key.length > 200 || m === 'GET' || m === 'HEAD' || m === 'OPTIONS') return next(req)
  const ct = req.headers.get('Content-Type')
  if (ct && !ct.toLowerCase().startsWith('application/json')) return next(req)
  if (!id.institution) return next(req)

  // The body is read to hash it; the handler gets a fresh Request over the same bytes.
  const body = new Uint8Array(await req.arrayBuffer())
  if (body.byteLength > 8 << 20) throw new HttpError(400, 'could not read the request body', { code: 'bad_request' })
  const inner = new Request(req, { body: m === 'GET' || m === 'HEAD' ? undefined : body })
  const hash = [...new Uint8Array(await crypto.subtle.digest('SHA-256', body))]
    .map((b) => b.toString(16).padStart(2, '0')).join('')

  const d = db()
  const inst = id.institution.id
  const path = new URL(req.url).pathname
  // A claim nobody settled within two minutes is wreckage from a killed request.
  const stale = new Date(Date.now() - 2 * 60_000).toISOString()
  const [, claim] = await d.batch([
    d.prepare(`DELETE FROM idempotency_keys WHERE institution_id = ? AND key = ?
                 AND completed_at IS NULL AND created_at < ?`).bind(inst, key, stale),
    d.prepare(`INSERT INTO idempotency_keys (institution_id, key, user_id, method, path, request_hash, created_at)
               VALUES (?, ?, ?, ?, ?, ?, ?) ON CONFLICT (institution_id, key) DO NOTHING`)
      .bind(inst, key, id.userId, m, path, hash, now()),
  ])

  if (!claim.meta.changes) {
    const prev = await d.prepare(`SELECT status_code, response_body, method, path, request_hash
                                    FROM idempotency_keys WHERE institution_id = ? AND key = ? AND user_id = ?`)
      .bind(inst, key, id.userId)
      .first<{ status_code: number | null; response_body: unknown; method: string; path: string; request_hash: string }>()
    if (!prev) throw new HttpError(409, 'That request reference belongs to a different sign-in. Try the action again.', { code: 'idempotency_key_reused' })
    if (prev.method !== m || prev.path !== path || prev.request_hash !== hash)
      throw new HttpError(409, 'That request reference was already used for a different request.', { code: 'idempotency_key_conflict' })
    if (prev.status_code == null)
      throw new HttpError(409, 'That request is still being processed. It will not be sent twice.', { code: 'idempotency_in_flight' })
    return new Response(toBytes(prev.response_body), {
      status: prev.status_code,
      headers: { 'content-type': 'application/json', 'Idempotent-Replay': 'true' },
    })
  }

  let res: Response
  try { res = await next(inner) } catch (err) { res = errorResponse(err) }
  const out = new Uint8Array(await res.arrayBuffer())

  // Settling must not fail the response the handler already produced.
  try {
    if (res.status >= 500) {
      await d.prepare(`DELETE FROM idempotency_keys WHERE institution_id = ? AND key = ?`).bind(inst, key).run()
    } else {
      await d.prepare(`UPDATE idempotency_keys SET status_code = ?, response_body = ?, completed_at = ?
                        WHERE institution_id = ? AND key = ?`).bind(res.status, out, now(), inst, key).run()
    }
  } catch (err) { console.error('idempotency settle', err) }
  return new Response(res.status === 204 || res.status === 304 ? null : out, { status: res.status, statusText: res.statusText, headers: res.headers })
}

function toBytes(v: unknown): Uint8Array {
  if (v == null) return new Uint8Array()
  if (v instanceof ArrayBuffer) return new Uint8Array(v)
  if (ArrayBuffer.isView(v)) return new Uint8Array(v.buffer, v.byteOffset, v.byteLength)
  if (Array.isArray(v)) return Uint8Array.from(v as number[])
  return new TextEncoder().encode(String(v))
}
